File: src/algorithms/python/Range_Sum_Query_Mutable.py
# -*- coding: utf-8 -*-


class NumArray(object):
    def __init__(self, nums):
        """
        initialize your data structure here.
        :type nums: List[int]
        """
        self.sums = [0 for i in range((len(nums) << 2) + 1)]
        self.n = len(nums)
        for index, num in enumerate(nums):
            self.SGUpdate(index + 1, num, 1, self.n, 1)

    def SGPushUP(self, rt):
        self.sums[rt] = self.sums[rt << 1] + self.sums[rt << 1 | 1]

    def SGUpdate(self, pos, c, l, r, rt):
        if l == r and l == pos:
            self.sums[rt] = c
            return
        m = (l + r) >> 1
        if pos <= m:
            self.SGUpdate(pos, c, l, m, rt << 1)
        else:
            self.SGUpdate(pos, c, m + 1, r, rt << 1 | 1)
        self.SGPushUP(rt)

    def SGQuery(self, L, R, l, r, rt):
        if L <= l and r <= R:
            return self.sums[rt]
        m = (l + r) >> 1
        ret = 0
        if L <= m:
            ret += self.SGQuery(L, R, l, m, rt << 1)
        if m < R:
            ret += self.SGQuery(L, R, m + 1, r, rt << 1 | 1)
        return ret

    def update(self, i, val):
        """
        :type i: int
        :type val: int
        :rtype: int
        """
        self.SGUpdate(i + 1, val, 1, self.n, 1)

    def sumRange(self, i, j):
        """
        sum of elements nums[i..j], inclusive.
        :type i: int
        :type j: int
        :rtype: int
        """
        return self.SGQuery(i + 1, j + 1, 1, self.n, 1)


if __name__ == "__main__":
    # Your NumArray object will be instantiated and called as such:
    nums = [1, 3, 5]
    numArray = NumArray(nums)
    print numArray.sumRange(0, 2)
    numArray.update(1, 2)
    print numArray.sumRange(0, 2)

File: src/algorithms/cpp/Range_Sum_Query_Mutable.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}

class NumArray {
private:
    vector <int> sums;
    int n;

public:
    NumArray(vector<int> &nums) {
        int len = (SZ(nums) << 2) + 1;
        for (int i = 0; i < len; ++ i)
            sums.push_back(0);
        this->n = SZ(nums);
        for (int i = 0; i < SZ(nums); ++ i) {
            this->SGUpdate(i + 1, nums[i], 1, this->n, 1);
        }
    }
    
    void SGPushUp(int rt) {
        this->sums[rt] = this->sums[rt << 1] + this->sums[rt << 1 | 1];
    }
    
    void SGUpdate(int pos, int c, int l, int r, int rt) {
        if (l == r && l == pos) {
            this->sums[rt] = c;
            return;
        }
        int mid = (l + r) >> 1;
        if (pos <= mid)
            this->SGUpdate(pos, c, l, mid, rt << 1);
        else
            this->SGUpdate(pos, c, mid + 1, r, rt << 1 | 1);
        this->SGPushUp(rt);
    }
    
    int SGQuery(int L, int R, int l, int r, int rt) {
        if (L <= l && r <= R) {
            return this->sums[rt];
        }
        int mid = (l + r) >> 1, ret = 0;
        if (L <= mid)
            ret += this->SGQuery(L, R, l, mid, rt << 1);
        if (mid < R)
            ret += this->SGQuery(L, R, mid + 1, r, rt << 1 | 1);
        return ret;
    }

    void update(int i, int val) {
        this->SGUpdate(i + 1, val, 1, this->n, 1);
    }

    int sumRange(int i, int j) {
        return this->SGQuery(i + 1, j + 1, 1, this->n, 1);
    }
};


int main() {
    return 0;
}
